src/env.h

```diff
@@ -263,6 +265,9 @@ class Environment {
     enum Fields {
       // Set this to not zero if the init hook should be called.
       kCallInitHook,
+      // Set this to not zero if async wrap should be enabled. This is
+      // automatically set when SetupHooks is called.
+      kEnabled,
```

Comment:
>  I know you know this, but to reiterate. The point was that before/after don't run unless init runs, so enabling init implies enabling the others.

Thanks for reiterating, I actually didn't realize that. The comment in (https://github.com/iojs/io.js/blob/v1.x/src/async-wrap.cc#L39) is rather confusing then:
> This only affects the init callback.

I realize now that it also says: 
> If the init callback was called, then the pre/post callbacks will automatically be called.

which should be read as "if and only if". Sometimes I think my math education have kills all implicit human understanding :)

In any case I find it a bit confusing, the `kCallInitHook` should really just be called `kEnable` then. Or is it because `before` and `after` callbacks can still be called after `kEnable` is called?

I would propose having a `kEnable` which enables (`1`) or short circuit (`0`) the `async_wrap` stuff. Such that some or no callbacks are called.

I'm not sure how much of a performance hit a no-opt call is these days, but I guess having  `kCallInitHook`, `kCallBeforeHook` and `kCallAfterHook` flags would make sense. The flags are such that the respective callback is only called the flag is set. For instance I don't think there is any reason for me to have an `after` hook in [`trace (L80)`](https://github.com/AndreasMadsen/trace/blob/master/trace.js#L80)

The logic behind preventing the `before` and `after` hook calls if `init` wasn't called, is then removed. I don't think this is particularly useful, it seams like it is user protection and it is easy to implement in userland.